Add support for adding single items to STAC collections

denarch22 opened the pull request:
## 🛠️ Changes
Introduces the 'add-item' command to add individual items to existing STAC collections on the server. Updates documentation, environment variables, and configuration to support asset base URLs. Implements item validation against collection specifications (projection, resolution, data type) and provides logic for processing, converting, uploading, and optionally deleting local COG files after upload.

## 📝 Associated issues
Resolve [LIB-328](https://linear.app/biodev/issue/LIB-328/comando-para-agregar-item-a-coleccion-existente)

## ✅ Checks
- [X] Verify that the docs has been updated according to the changes in this PR

denarch22 committed:
Add support for adding single items to STAC collections

Introduces the 'add-item' command to add individual items to existing STAC collections on the server. Updates documentation, environment variables, and configuration to support asset base URLs. Implements item validation against collection specifications (projection, resolution, data type) and provides logic for processing, converting, uploading, and optionally deleting local COG files after upload.

denarch22 committed:
Merge branch 'darmenta/LIB-387' into darmenta/LIB-328

cgalvist reviewed (changes requested):
Hola.

No he revisado la funcionalidad pero tengo estas observaciones por ahora:

1. No se ha actualizado la documentación de la página con `Github Pages`
2. Al ejecutar el comando `python src/format.py`, me cambia muchas líneas de 3 archivos

denarch22 committed:
Refactor add-item flow and improve validation logic

Refactored the add-item command flow in main.py for clarity and maintainability, including improved logging and error handling. Enhanced item validation logic in collection.py for projection, resolution, and data type checks. Improved YAML doc formatting and display name generation in gen_pages.py. Fixed projection EPSG validation in spec.py. Renamed add_item.yml to add item.yml for consistency.

cgalvist reviewed (changes requested):
Hola.
He realizado esta prueba:

1. Crear una colección nueva (copia de coberturas): `python src/main.py create -f test -c cgt_test`
  - Al crear la colección, me aparece solo un ítem `2020`, lo cuál está bien.
2. Creo una copia del archivo `CLC_clases_2020.tif` con el nombre `2020-2.tif` y ejecuto este comando para añadir un ítem nuevo: `python src/main.py add-item -c cgt_test -f test --file 2020-2.tif --year 2021`
  - Al revisar los ítems, me aparecen 2 con los ids `2020` y `2020-2`, que también está bien

Me queda complicado realizar pruebas con diferentes resoluciones y proyecciones, pero en general esos resultados están bien.

Para probar la validación de los tipos de colecciones (antes "binaria o continua", ahora "clasificada o continua"), necesito que porfa hagas merge de los cambios de main en tu rama para seguir haciendo pruebas

cgalvist reviewed (changes requested):
Todavía no veo los cambios de main en tu rama

denarch22 commented:
Listo ya entendí que  paso jejeje

ManuelStardust commented on src/collection.py:
```diff
@@ -221,6 +222,95 @@ def check_collection(self, overwritten):
+            "metadata" in collection_metadata
+            and "data_type" in collection_metadata.get("metadata", {})
+        ):
+            collection_metadata["metadata"]["data_type"]
```
Creo que esta parte dle código no esta haciendo nada o no me queda claro para que se usa.

ManuelStardust commented on docs/gen_pages.py:
```diff
@@ -19,11 +19,22 @@ def gen_files():
+            
+            display_name = name.replace("_", " ").title()
+            
+            import yaml
```
Los import deberian estar el inicio del archivo.

denarch22 committed:
Validate item dtype against collection data_type

Adds validation to ensure that the data type of new items matches the collection's 'Clasificada' (integer) or 'Continua' (float) data_type. Also moves the yaml import in docs/gen_pages.py to the top of the file for consistency.

denarch22 committed:
Merge branch 'main' into darmenta/LIB-328

cgalvist commented on src/collection.py:
```diff
@@ -221,6 +222,113 @@ def check_collection(self, overwritten):
+        ):
+            collection_data_type = collection_metadata["metadata"]["data_type"]
+            
+            # Validate that item dtype is compatible with collection data_type
```
Hola.
He realizado esta prueba otra vez:

1. Crear una colección nueva (copia de coberturas): `python src/main.py create -f test2`
  - Al crear la colección, me aparece solo un ítem `2020`, lo cuál está bien.
2. Creo una copia del archivo `CLC_clases_2020.tif` con el nombre `2020-2.tif` y ejecuto este comando para añadir un ítem nuevo: `python src/main.py add-item -c CesarTest2 -f test --file 2020-2.tif --year 2021`
  - Al ejecutar el comando, me aparece el error `ValueError: Data type mismatch: collection is 'Clasificada' (requires integer dtype), but item has dtype 'float32'`. Lo cuál considero que está bien, ya que no se deberían usar valores decimales en capas discretas.

Sin embargo, me queda la duda del comentario de @erikasv en el ticket [326](https://linear.app/biodev/issue/LIB-326/agregar-propiedad-para-identificar-si-coleccion-es-discreta-o-continua), en donde indica los tipos de numeración que se pueden usar en las capas (numérico y ordinal). 

En la explicación de @erikasv, entiendo que es posible que un ráster continuo tenga valores enteros, por lo que creo que esta función también mostraría un error.

Personalmente sí estoy de acuerdo con tus validaciones, y en su momento propuse usar una clasificación similar a la tuya en el ticket, pero no sé que hacer en este caso. Desde mi parte, aprobaría este ticket pero creo que debemos discutirlo con el equipo

denarch22 replied:
Hola @cgalvist ,
Gracias por la prueba y el comentario.

El error que aparece es correcto: una colección "Clasificada" no debe aceptar valores flotantes, ya que representa datos discretos/categóricos.
Sobre el punto de @erikasv en el ticket 326: es válido que un ráster continuo pueda tener valores enteros cuando representan datos numéricos u ordinales (por ejemplo, índices o valores discretos que representan rangos continuos).
La validación actual es estricta:
"Clasificada": solo acepta tipos enteros (uint8, uint16, uint32, int16, int32)
"Continua": solo acepta tipos flotantes (float32, float64)
Se podria ajustar la validación para que "Continua" acepte tanto enteros como flotantes, manteniendo "Clasificada" solo con enteros. 

Esto permitiría:
Rechazar flotantes en colecciones "Clasificada" (correcto)
Aceptar enteros en colecciones "Continua" (para casos numéricos/ordinales)
Aceptar flotantes en colecciones "Continua" (casos tradicionales)

¿Te parece bien esta propuesta? Si estás de acuerdo, la implemento, o si prefieren otra lógica, podemos discutirlo con el equipo antes de cambiar el código.

cgalvist replied:
Hola.

La verdad estoy de acuerdo en que sea estricto y yo lo dejaría así como está, pero tal vez @erikasv no lo esté 🙈.

Como te comenté, desde mi punto de vista estaría aprobado este PR, pero creo que es algo que debemos discutir con @erikasv . Tal vez lo mejor sea aprobar esto como está y hacer otro desarrollo después de ser necesario, ya que @ManuelStardust también lo aprobó.

denarch22 replied:
@cgalvist Creo que falta tu aprobación. jejeje

ManuelStardust replied:
En ese caso seria bueno dejar el resultado de esta discusión como comentario en linear.
